kip-0028.md

```diff
@@ -0,0 +1,174 @@
+  - the message ID `messageId` that is the keccak256 hash of `msg`:
+
+The **Hyperlane Message** verifier has as input `msg` and as output the singleton set containing the Pact capability `(B.MINT M signerPublicKeys messageId amt recipient)`. The bridge module is responsible for:
+  - registering `messageId` as received in its storage, preventing message replays;
+  - checking that signature by `signerPublicKeys` is sufficient to authenticate the message;
+  - calling into `M` to ensure that the mint happens.
+
+Conventionally, Hyperlane bridges are secured by "Interchain Security Modules" or ISMs, which both contain cryptographic code and manage the set of *validators*, that is, public keys that can authenticate Hyperlane messages. This verifier implements the *mechanism* of the ISM, that is the cryptography, while leaving the *policy* of the ISM up to the bridge code, that is, managing the set of validators and ensuring that the signers match the validators in some way.
+
+# Implementation
+
+## `chainweb-node`
+
+In `chainweb-node` we propose adding a mapping from verifier names to verifier functions, which is consulted when validating a transaction. A verifier may charge gas after gas is bought and before any transaction code is run. Because of this, currently, a verifier cannot grant the `GAS_PAYER` capability, which would allow paying for gas using a verifier’s attestation. This may change in the future.
+
+Notably, writing a verifier plugin does not require changes to `pact`, only to `chainweb-node`, because verifiers run on the Pact container layer and verifier names are strings.
+
+## Changes to the Pact transaction schema
+
+For reference, this is the schema of the `signers` field:
+
+`signers` is a required array of objects with:
+
+- `pubKey`:
+    - the public key image. Pact default is base16 ED25519 encoding.
+        - a string.
+- `clist`:
+    - a list of capabilities associated with/installed by this signer.
+        - an optional array of objects with:
+            - `name`: a fully-qualified capability name, as a string.
+            - `args`: an array of JSON-encoded Pact values.
+
+And an example:
+
+```json
+"signers":
+  [
+    {
+      "pubKey": "368820f80c324bbc7c2b0610688a7da43e39f91d118732671cd9c7500ff43cca",
+      "clist":
+        [
+          {"name": "coin.GAS", "args": []},
+          {"name": "coin.TRANSFER", "args": ["sender00", "sender01", 1]}
+        ]
+    }
+  ]
+
+```
+
+The public key is assumed to be an Ed25519 key by default. The signatures are
+located elsewhere in the transaction, but they are assumed to correspond with
+signers. The first signature corresponds to the first signer, second signature
+to the second signer, and so on.
+
+We propose the following schema for including extra verification types in
+transactions:
+
+`verifiers`, an array of objects with:
+
+- `name`:
+    - an identifier for the verification type.
+        - a string.
+- `proof`:
```

Comment:
Is the proof a string or JSON? Since the example is a string, but here it says JSON

Reply:
I think, that the proof string just be just binary. In the JSON body of the tx it would be encoded as base64url (without padding).

Proof strings messages from the outside world that don't have a meaning in Pact. We should constrain the content as little as possible. So plain bytes seem the right choice. 

It is the job of the verifier to understand those plain bytes and relate them to the Pact semantics of the content of the `clist`.

Reply:
The proof string right now is indeed a JSON representation of a Pact value. We can change this later in a subsequent KIP.